selfdrive/locationd/lagd.py

```diff
@@ -0,0 +1,287 @@
+      self.valid_blocks = min(self.valid_blocks + 1, self.num_blocks)
+
+  def get(self):
+    valid_block_idx = [i for i in range(self.valid_blocks) if i != self.block_idx]
+    if not valid_block_idx:
+      return None
+    return np.mean(self.values[valid_block_idx], axis=0)
+
+
+class LagEstimator:
+  def __init__(self, CP, dt,
+               block_count=BLOCK_NUM, block_size=BLOCK_SIZE,
+               window_sec=MOVING_WINDOW_SEC, okay_window_sec=MIN_OKAY_WINDOW_SEC,
+               min_vego=MIN_VEGO, min_yr=MIN_ABS_YAW_RATE, min_ncc=MIN_NCC):
+    self.dt = dt
+    self.window_sec = window_sec
+    self.okay_window_sec = okay_window_sec
+    self.initial_lag = CP.steerActuatorDelay + 0.2
+    self.block_size = block_size
+    self.block_count = block_count
+    self.min_vego = min_vego
+    self.min_yr = min_yr
+    self.min_ncc = min_ncc
+
+    self.t = 0
+    self.lat_active = False
+    self.steering_pressed = False
+    self.steering_saturated = False
+    self.desired_curvature = 0
+    self.v_ego = 0
+    self.yaw_rate = 0
+
+    self.calibrator = PoseCalibrator()
+
+    self.reset(self.initial_lag, 0)
+
+  def reset(self, initial_lag, valid_blocks):
+    window_len = int(self.window_sec / self.dt)
+    self.points = Points(window_len)
+    self.block_avg = BlockAverage(self.block_count, self.block_size, valid_blocks, initial_lag)
+    self.lag = initial_lag
+
+  def get_msg(self, valid):
+    msg = messaging.new_message('liveActuatorDelay')
+
+    msg.valid = valid
+
+    liveDelay = msg.liveActuatorDelay
+    liveDelay.steerActuatorDelay = self.lag
+    liveDelay.isEstimated = self.block_avg.valid_blocks > 0
+    liveDelay.validBlocks = self.block_avg.valid_blocks
+    liveDelay.points = self.block_avg.values.tolist()
+
+    return msg
+
+  def handle_log(self, t, which, msg):
+    if which == "carControl":
+      self.lat_active = msg.latActive
+    elif which == "carState":
+      self.steering_pressed = msg.steeringPressed
+      self.v_ego = msg.vEgo
+    elif which == "controlsState":
+      self.desired_curvature = msg.desiredCurvature
+      self.steering_saturated = getattr(msg.lateralControlState, msg.lateralControlState.which()).saturated
+    elif which == "liveCalibration":
+      self.calibrator.feed_live_calib(msg)
+    elif which == "livePose":
+      device_pose = Pose.from_live_pose(msg)
+      calibrated_pose = self.calibrator.build_calibrated_pose(device_pose)
+      self.yaw_rate = calibrated_pose.angular_velocity.z
+    self.t = t
+
+  def points_valid(self):
+    return self.points.num_okay >= int(self.okay_window_sec / self.dt)
+
+  def update_points(self):
+    la_desired = self.desired_curvature * self.v_ego * self.v_ego
+    la_actual_pose = self.yaw_rate * self.v_ego
+
+    fast = self.v_ego > self.min_vego
+    turning = np.abs(self.yaw_rate) >= self.min_yr
+    okay = self.lat_active and not self.steering_pressed and not self.steering_saturated and fast and turning
```

Comment:
You should check that the user hasn't been steering and we've been engaged for longer than some time frame since there is some ramp time back to a steady control state.

https://github.com/commaai/openpilot/blob/a2859090d835041a0c662809f7549bdc2040af16/selfdrive/locationd/torqued.py#L196-L198

Reply:
yeah, thats planned (the 4th point of the todo list)

Reply:
implemented in [d00f784](https://github.com/commaai/openpilot/pull/34531/commits/d00f7841d5668efb54afd0af7b8e37615c988505)